Auth/Auth/AbstractAuthService.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Threading.Tasks;
+using Aerogear.Mobile.Auth.User;
+using AeroGear.Mobile.Auth.Authenticator;
+using AeroGear.Mobile.Auth.Config;
+using AeroGear.Mobile.Auth.Credentials;
+using AeroGear.Mobile.Core;
+using AeroGear.Mobile.Core.Configuration;
+using static AeroGear.Mobile.Core.Utils.SanityCheck;
+
+namespace AeroGear.Mobile.Auth
+{
+    /// <summary>
+    /// Abstract implementation of <see cref="IAuthService"/>.
+    /// </summary>
+    public abstract class AbstractAuthService : IAuthService
+    {
+        protected CredentialManager CredentialManager { get; set; }
+        protected KeycloakConfig KeycloakConfig { get; set; }
+        protected AuthenticationConfig AuthenticationConfig { get; private set; }
+        protected MobileCore MobileCore { get; set; }
+        protected IAuthenticator Authenticator { get; set; }
+        public string Type => "keycloak";
+        public bool RequiresConfiguration => true;
+
+        public abstract User CurrentUser();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:AeroGear.Mobile.Auth.AbstractAuthService"/> class.
+        /// </summary>
+        /// <param name="mobileCore">Mobile core.</param>
+        /// <param name="serviceConfig">Service config.</param>
+        public AbstractAuthService(MobileCore mobileCore = null, ServiceConfiguration serviceConfig = null)
+        {
+            MobileCore = mobileCore ?? MobileCore.Instance;
+            var serviceConfiguration = NonNull(serviceConfig ?? mobileCore.GetServiceConfiguration(Type), "serviceConfig");
+            KeycloakConfig = new KeycloakConfig(serviceConfiguration);
+        }
+
+        /// <summary>
+        /// Initiate an authentication flow.
+        /// </summary>
+        /// <returns>The authenticate.</returns>
+        /// <param name="authenticateOptions">Authenticate options.</param>
+        public Task<User> Authenticate(IAuthenticateOptions authenticateOptions)
+        {
+            return Authenticator.Authenticate(authenticateOptions);
```

Comment:
This and logout and everything should ensure everything is configured.

I guess the `Configure` function could go away though if we provide `AuthenticationConfig` to `InitializeService` and then provide it to the constructor. That would be simpler.